### PR TITLE
Issue 66496

### LINKED ISSUE #64335: Generic attributes: Attribute.GetCustomAttributes returns null when passed an unbound generic type
### Description

`Attribute.GetCustomAttributes` normally returns an empty array when no attributes are found (or throws an exception if the parameters don't make sense).

However,  `Attribute.GetCustomAttributes(someType, typeof(SomeGenericAttribute<>))` returns `null`. This is unexpected, particularly as the nullable annotations say that it doesn't return `null`.

It's worth noting that this wasn't previously possible: it was simply not possible to create a generic type which derived from `Attribute`, and if you passed a type which wasn't derived from `Attribute` to `Attribute.GetCustomAttributes`, you got an exception telling you that. So I suspect this is an unhandled case which just fell through the cracks.

### Reproduction Steps

```cs
Test.M();

public class Test
{
    public static void M()
    {
        var attrs = Attribute.GetCustomAttributes(typeof(Test), typeof(TestAttribute<>));
        Console.WriteLine(attrs == null); // true
    }
}

public class TestAttribute<T> : Attribute { }
```

[SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQHMIDsIIJYDGqsM+wcMKANDCAnFlQCYgDUAPgAIBMAjALAAoIQBUUMAHQBZABQBKANxChnAMwACHurHIYQgN5D1xzRs68AbJoAs62XKMnDgk6/UA3KAnUkEydQC86gCCMKR45JQSAOIQMADCcLoA9gC2oeGRKDIwAJ4ADhDJYDI6MHJU6nmFxaXiGWQUEAA8AHxyio5umrwAnDK+/gFBWHAANmOK6gD001X0EF3qAL5Cq8KCaprc2vVhjZTNIq3qICH7EU3q+itAA===)

### Expected behavior

`Attribute.GetCustomAttributes` either returns an array, or throws an exception. It should not return `null`.

### Actual behavior

`Attribute.GetCustomAttributes` returns `null` in this case.

### Regression?

No. It was not previously possible to pass an unbound generic attribute type to `Attribute.GetCustomAttributes`.

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by madelson**
Restore the reverted PR (https://github.com/dotnet/runtime/pull/65237/files) but this time handle DynamicMethod
Fixes [66496 ](https://github.com/dotnet/runtime/issues/64335)

### CODE REVIEW

**Comment by jkotas** [comment]
The new test is failing on Mono:
```
[02:46:22] fail: [FAIL] System.Tests.AttributeGetCustomAttributes.GetCustomAttributes_DynamicMethod
[02:46:22] info: Assert.Equal() Failure
[02:46:22] info: Expected: 1
[02:46:22] info: Actual:   0
[02:46:22] info:    at System.Tests.AttributeGetCustomAttributes.GetCustomAttributes_DynamicMethod()
[02:46:22] info:    at System.Reflection.RuntimeMethodInfo.InvokeWorker(Object obj, BindingFlags invokeAttr, Span`1 parameters)
[02:48:28] info: Finished:    System.Runtime.Tests.dll
```

**Comment by madelson** [comment]
> The new test is failing on Mono:

@jkotas I spent some time looking into this and it seems to me that there is a bit of a web to untangle. TL;DR I set the test to skip on Mono.

First off, unlike with coreclr Mono's attempt to support `MethodImplAttribute` in `DynamicMethod` does not extend to the compiled method. That means you get a different answer when you call `GetCustomAttributes()` on the original `DynamicMethod` instance vs. the `Method` property of the created `Delegate`. On CoreCLR these call through to the same code in `RTDynamicMethod`. To solve for this I think we'd need Mono start using something like `RTDynamicMethod`.

If we restrict the problem to just looking at calling `Attribute.GetCustomAttributes()` on the `DynamicMethod` instance directly, there is a separate issue. Because `CustomAttribute.IsUserCattrProvider` thinks `DynamicMethod` should be treated as "native", it won't actually call through to `DynamicMethod.GetCustomAttributes()`. This is easy enough to change with a tweak to `CustomAttribute.IsUserCattrProvider`, but then we run into the issue that `CustomAttribute.GetCustomAttributesBase()` ends up calling the underlying `GetCustomAttributes()` passing `null` for the attribute type when the type starts out as `typeof(Attribute)` or is not provided. There are multiple `FIXME` comments in the file warning about this. I don't think this would be a huge deal to fix, but the cleanest way I can see to do it (using `typeof(object)` as the sole means to request unfiltered attributes) would result in a good number of changes to `CustomAttribute.cs` and `Attribute.Mono.cs`. Without more context, it's hard for me to say whether those changes might have other unforseen consequences.

In short, fixing either of these felt out of scope for the original problem; from what I can tell these are gaps/oddities in the Mono attribute functionality that are preexisting and not related to generic attributes at all.

Thoughts?

**Comment by jkotas** [comment]
I agree with your assessment.

**Review by jkotas — APPROVED** [review]

File: src/tests/reflection/GenericAttribute/GenericAttributeTests.cs [hunk 1]
```diff
@@ -18,9 +18,17 @@ static int Main(string[] args)
...
+        Assert(!CustomAttributeExtensions.GetCustomAttributes(assembly, typeof(SingleAttribute<>)).GetEnumerator().MoveNext());
 */
 
+        // Module module = programTypeInfo.Module;
+        // AssertAny(CustomAttributeExtensions.GetCustomAttributes(module), a => a is SingleAttribute<long>);
+        // Assert(CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<long>)).GetEnumerator().MoveNext());
+        // Assert(CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<long>)).GetEnumerator().MoveNext());
+        // Assert(!CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<>)).GetEnumerator().MoveNext());
+        // Assert(!CustomAttributeExtensions.GetCustomAttributes(module, typeof(SingleAttribute<>)).GetEnumerator().MoveNext());	
+
```
Comment (deeprobin) [inline src/tests/reflection/GenericAttribute/GenericAttributeTests.cs:31]: Can we remove these comments?

**Review by deeprobin — APPROVED** [review]
LGTM

File: src/tests/reflection/GenericAttribute/GenericAttributeTests.cs [hunk 2]
```diff
@@ -18,9 +18,17 @@ static int Main(string[] args)
         Assert(((ICustomAttributeProvider)assembly).IsDefined(typeof(SingleAttribute<int>), true));
         Assert(CustomAttributeExtensions.IsDefined(assembly, typeof(SingleAttribute<bool>)));
         Assert(((ICustomAttributeProvider)assembly).IsDefined(typeof(SingleAttribute<bool>), true));
-
+        Assert(!CustomAttributeExtensions.GetCustomAttributes(assembly, typeof(SingleAttribute<>)).GetEnumerator().MoveNext());
+        Assert(!CustomAttributeExtensions.GetCustomAttributes(assembly, typeof(SingleAttribute<>)).GetEnumerator().MoveNext());
 */
 
+        // Module module = programTypeInfo.Module;
```
Comment (buyaa-n) [inline src/tests/reflection/GenericAttribute/GenericAttributeTests.cs:25]:
```suggestion
        // Uncomment when https://github.com/dotnet/runtime/issues/66168 is resolved
        // Module module = programTypeInfo.Module;
```

File: src/tests/reflection/GenericAttribute/GenericAttributeTests.cs (hunk 1 above)
Comment (buyaa-n) [reply src/tests/reflection/GenericAttribute/GenericAttributeTests.cs]: See https://github.com/dotnet/runtime/issues/66168

**Review by buyaa-n — APPROVED** [review]
LGTM, thanks!

**Review by marek-safar — APPROVED** [review]

**Comment by buyaa-n** [comment]
Build failures unrelated and [known issue](https://github.com/dotnet/runtime/issues/66556), merging
